### DISCUSSION TITLE
CDM Design Principles

### DISCUSSION THREAD

**Discussion by tomhealey-icma** [finos/common-domain-model#2422] [General]
With due consideration given to the Design Guidelines, https://cdm.docs.rosetta-technology.io/source/contribution.html, this section is intended to establish a set of domain design principles and sample scenarios to help guide contributors.  The major sections of the design principles should include:

- CDM Subdomains and Scopes
- Type and Attribute Naming Conventions
- Guidelines on Creating New Data Types
- Guidelines on Modifying Existing Data Types
- Separation of Workflow and Business Events

**CDM Subdomain and Scopes**

Types identify distinct entities and should be designed based on the MECE Principle. As such, types representing products or other classes should be mutually exclusive, meaning that no other Type in the same category can mutually exist. The second part of the MECE Principle states that entities should be collectively exhaustive. This principle means that an entity should fully encompass all elements that may exists exclusive of other entities.

The CDM core model should not include any specific type or attribute that identifies it with any regulation, organization, messaging system or vendor. Instead domains outside the core CDM should be associated through a proper class definition that broadly covers all subclasses and identities within that domain. Specifically, the use of FIX, FpML, ISDA, ICMA, ISLA, SWIFT, SFTR, MiFID should not be used within the core model, either structurally or in name.

**Type and Attribute Naming Conventions**

Enumerations

Enumerations should be in all upper case with no special characters except the underscore '_'. Enumeration phrases should separate words using the _ instead of running the words together. 

Acceptable prefix and suffix phrases

The use of prefix and suffix phrases should be consistent in the definition across all primary data types where they may be used to define a subtype or modify an data type or attribute. 

As an example, the use of Provision to modify a Termination refers specifically to a complex data type that contains attributes that are needed to properly define and execute an termination. Other phrases such as List, Schedule, Method would not fully capture the purpose. Prefix modifiers create a subclass of the primary data type. In this case, TerminationDate redefines Date. This approach creates clarity in the model as a opposed to using a Termination type with an attribute that defines it as a Termination Date and results in both data and structural consistency in the model.

Proposed list of primary data type modifiers:

Adjustment
Method
Provision
Calculation
Schedule
List
Terms
Specification
Representation
Reference
Choice
Range
Information

**Comment by brianlynn2** [comment]
The above guidelines are a good start on coding and low level design guidelines.  I think that there need to be some higher level principles/guidelines on how the overall design should evolve, to improve supportability over time.  I've appended my ideas on that below.  In addition, we should have guidelines on how the design approach should work, e.g  reuse of existing types vs specialization for differing requirements.  This might fall under the category of "creating new types" above.

### CDM Supportability Goals and Guidelines


**Supportability Goals**

To support CDM adoption and use, we have the following supportability goals for CDM.

- **Stable production support:**  Firms developing functionality based on CDM should be able to continue to use that functionality over an extended period of time with relatively little to no maintenance effort (e.g. without deploying new versions of CDM), as long as their functionality needs do not change.  Any reference data used by CDM should be able to be updated without maintenance to the software itself.

_**Rationale:**  Financial systems that meet their mission can remain unchanged (except for reference data) for long periods of time, sometime over a decade.  Firms developing based on CDM will not wish to invest resources in tracking new versions of CDM for functionality that has not changed._

- **Migration to new CDM versions:**  Firms wishing to adopt newer versions of CDM should be able to migrate to that new functionality with minimal effort.

_**Rationale:**  Developers using CDM want to protect their investment to continue to adopt new versions of CDM as CDM changes, otherwise they are not as likely to  choose to build based on CDM._

- **Interoperability between CDM versions:**  It should be possible for firms deploying systems based on different versions of CDM to interoperate, at least over a defined version range. As we expect CDM-based systems to be long-lived, ideally we would desire new CDM-based systems to be able to interoperate with other CDM-based systems developed a long time ago (years to over a decade in the past).  

_**Rationale:**  Once financial systems are put into place, it is common for them to stay in production for a long time.  Firms deploying those systems are reluctant to invest money in adapting to new versions of underlying software and standards unless there is an urgent business need to do so, so underlying technology tends to change relatively slowly.   We know of FpML-based systems that have been in production for close to 20 years, in some cases with little or no change to the core logic.   Ideally it shouldn’t be necessary to change a CDM-based system that fulfills its mission just to interoperate with another CDM-based system that is based on a newer version of CDM._
### 

**High Level Design Guidelines for Supportability**

The following guidelines are intended to support the design goals above.  (Note that the existing CDM design does not meet all of the below guidelines.)

- **Reference Data:**  CDM-based applications based on any production version of CDM shall be able to continue to function without update to core logic as reference data is updated.  This implies that reference data shall be stored, updated, and distributed separately from code (data models and functions)  in a format or formats that remain(s) stable over a long period of time.   Ideally the reference data will be kept in data rather than code form, so that the data can be updated without a new code deployment, and to simplify reference data updates. 

_**Rationale:**  It should not be necessary to adapt or regression test existing functionality against new CDM functionality just to take advantage of new reference data.   For example, FpML reference lists have historically (over the last decade) changed around monthly, over 140 times in 11 years.   (And each change typically affects numerous data values, often in several lists.)  Production applications will need to be able to use that new data in a timely fashion without using new versions of CDM functionality, which may introduce new problems.  Typically the time between definition of new reference data and its use in production is a few weeks or less._

- **Change control guidelines:**  Developers of CDM-based applications shall have confidence that they will be able to update to new versions of CDM without major rework to their applications.  This shall be achieved as follows:

- [ ] Functionality in the CDM shall not be removed without approval of the Steering WG (SWG), and only then with confirmation from the industry that this functionality is not in active use
- [ ] An effort shall be made to minimize the changes or deletions of data fields and changes to type names for any data fields present in a production release of CDM.   For SWG designated “stable” types (such as trades/products/streams) if any change is made to data field names or type names, a backward-compatibility feature shall be added to map old data formats to new data formats unless the SWG agrees that this functionality is not in active use and explicitly approves breaking compatibility.  This compatibility shall be provided for both code-based interfaces (calls by users to CDM data types and functions)  and for JSON-persisted data.  (TBD:  should any CDM functions, e.g. top-level functions,  be designated as “stable” for code interfacing purposes?)
- [ ] New versions of the CDM shall be regression tested against applications using legacy interfaces and legacy data formats to ensure that the above requirements are met.

- **Interoperability Guidelines:**  Mechanisms shall be provided to allow newer versions of CDM to interoperate with older versions of CDM that remain supported for interoperability purposes..

- [ ] Newer versions of CDM shall be able to read serialized “stable” CDM objects produced in older supported versions of CDM.
- [ ] Newer versions of CDM shall be able to serialize the “stable” objects in the older CDM format, subject to the limitation that any new data fields cannot be serialized and would be omitted, so there may be loss of data related to new features.


- **Role of the SWG:**  The Steering WG shall be responsible for designating:

- [ ] Which CDM versions are supported for reference data maintenance purposes.
- [ ] Which CDM versions are intended to be interoperable for data transmission purposes.
- [ ] Which CDM objects are considered “stable” for supportability/interoperability purposes.

- **Testing/Verification:** Artifacts based on old versions of CDM (e.g. demo applications and persisted data) shall be maintained and new CDM releases shall be tested against them during the release process to confirm that the above guidelines have been met.

**Comment by lolabeis** [comment]
As mentioned in yesterday's TAG #2360, I suggest moving this discussion to being... a `Discussion` in GitHub - Any objection?
